hibernate-core/src/main/java/org/hibernate/id/insert/GetGeneratedKeysDelegate.java

```diff
@@ -21,95 +21,110 @@
-		return insert;
+	public GetGeneratedKeysDelegate(
+			EntityPersister persister,
+			boolean inferredKeys,
+			EventType timing) {
+		super( persister, timing, inferredKeys );
+		this.inferredKeys = inferredKeys;
+
+		if ( inferredKeys ) {
+			columnNames = null;
+		}
+		else {
+			final List<GeneratedValueBasicResultBuilder> resultBuilders = jdbcValuesMappingProducer.getResultBuilders();
+			final List<String> columnNamesList = new ArrayList<>( resultBuilders.size() );
+			final boolean unquote = dialect().unquoteGetGeneratedKeys();
+			for ( GeneratedValueBasicResultBuilder resultBuilder : resultBuilders ) {
+				final String columnName = resultBuilder.getModelPart().getSelectionExpression();
+				columnNamesList.add( unquote ? unquote( columnName, dialect() ) : columnName );
+			}
+			columnNames = columnNamesList.toArray( new String[0] );
+		}
 	}
 
 	@Override
-	public TableInsertBuilder createTableInsertBuilder(
-			BasicEntityIdentifierMapping identifierMapping,
+	public TableMutationBuilder<?> createTableMutationBuilder(
 			Expectation expectation,
 			SessionFactoryImplementor factory) {
-		final TableInsertBuilder builder =
-				new TableInsertBuilderStandard( persister, persister.getIdentifierTableMapping(), factory );
-
-		final OnExecutionGenerator generator = (OnExecutionGenerator) persister.getGenerator();
-		if ( generator.referenceColumnsInSql( dialect ) ) {
-			final String[] columnNames = persister.getRootTableKeyColumnNames();
-			final String[] columnValues = generator.getReferencedColumnValues( dialect );
-			if ( columnValues.length != columnNames.length ) {
-				throw new MappingException("wrong number of generated columns");
-			}
-			for ( int i = 0; i < columnValues.length; i++ ) {
-				builder.addKeyColumn( columnNames[i], columnValues[i], identifierMapping.getJdbcMapping() );
-			}
+		if ( getTiming() == EventType.INSERT ) {
+			return new TableInsertBuilderStandard( persister, persister.getIdentifierTableMapping(), factory );
+		}
+		else {
+			return new TableUpdateBuilderStandard<>( persister, persister.getIdentifierTableMapping(), factory );
 		}
-
-		return builder;
 	}
 
 	@Override
-	public PreparedStatement prepareStatement(String insertSql, SharedSessionContractImplementor session) {
+	public PreparedStatement prepareStatement(String sql, SharedSessionContractImplementor session) {
 		MutationStatementPreparer preparer = session.getJdbcCoordinator().getMutationStatementPreparer();
-		return inferredKeys
-				? preparer.prepareStatement( insertSql, RETURN_GENERATED_KEYS )
-				: preparer.prepareStatement( insertSql, persister.getRootTableKeyColumnNames() );
+		return columnNames == null
+				? preparer.prepareStatement( sql, RETURN_GENERATED_KEYS )
+				: preparer.prepareStatement( sql, columnNames );
 	}
 
 	@Override
-	public Object performInsert(
-			PreparedStatementDetails insertStatementDetails,
+	public boolean supportsArbitraryValues() {
+		return !inferredKeys;
```

Comment:
You can remove the inferredKeys field if you do this instead:

```suggestion
		return columnNames != null;
```